File: chapters/02_channel_adapter_registry/v2_rule_selection.py
from __future__ import annotations


DEFAULT_RULES = [
    {"when": {"channel": "telegram"}, "adapter": "telegram_adapter"},
    {"when": {"channel": "discord"}, "adapter": "discord_adapter"},
    {"when": {"channel": "webchat"}, "adapter": "webchat_adapter"},
]


def choose_adapter(envelope: dict[str, str], rules: list[dict[str, object]] | None = None) -> str:
    use_rules = rules or DEFAULT_RULES
    for rule in use_rules:
        when = rule.get("when", {})
        if not isinstance(when, dict):
            continue
        if all(envelope.get(k) == v for k, v in when.items()):
            return str(rule.get("adapter", "default_adapter"))
    return "default_adapter"


def dispatch_with_rules(envelope: dict[str, str]) -> str:
    adapter = choose_adapter(envelope)
    return f"{adapter}:handled:{envelope.get('text', '')}"


def main() -> None:
    env = {"channel": "discord", "text": "ping"}
    print(dispatch_with_rules(env))


if __name__ == "__main__":
    main()

File: chapters/02_channel_adapter_registry/v2_rule_selection.ts
// TypeScript mirror for Chapter 02: Channel Adapter Registry (v2)
// Run with: npx ts-node v2_rule_selection.ts

interface Rule {
  when: Record<string, string>;
  adapter: string;
}

const DEFAULT_RULES: Rule[] = [
  { when: { channel: "telegram" }, adapter: "telegram_adapter" },
  { when: { channel: "discord" }, adapter: "discord_adapter" },
  { when: { channel: "webchat" }, adapter: "webchat_adapter" },
];

function chooseAdapter(
  envelope: Record<string, string>,
  rules?: Rule[]
): string {
  const useRules = rules ?? DEFAULT_RULES;

  for (const rule of useRules) {
    const when = rule.when;
    const matches = Object.entries(when).every(
      ([k, v]) => envelope[k] === v
    );
    if (matches) return rule.adapter;
  }
  return "default_adapter";
}

function dispatchWithRules(envelope: Record<string, string>): string {
  const adapter = chooseAdapter(envelope);
  return `${adapter}:handled:${envelope.text ?? ""}`;
}

function main(): void {
  const env = { channel: "discord", text: "ping" };
  console.log(dispatchWithRules(env));
}

main();
